src/coap_gnutls.c

```diff
@@ -1432,6 +1465,8 @@ coap_dtls_hello(coap_session_t *c_session,
   ret = do_gnutls_handshake(c_session, g_env);
   if (ret == 1)
     return 1;
+  if (g_env->seen_client_hello)
```

Comment:
I wonder what happens when the handshake aborts with an error (i.e., `ret == -1`). It looks as if there might be cases where `seen_client_hello` is still 1. Is returning 1 in this case correct?

(The same holds for line 1456 f.)

Reply:
This is all down to needing to set up a new session hanging off endpoint->sessions earlier in the SSL handshake process.

Once we have passed the client_hello stage in the handshake process, GnuTLS then starts populating the gnutls_session_t with the (mismatched) keys currently being used etc. etc.   If endpoint->hello continues to hold the information that points to the  (now populated/polluted) gnutls_session_t, any new session coming in will fail as GnuTLS thinks this is a resume of an old session.

By trapping at the point when the client hello comes in and setting seen_client_hello, we still have a clean gnutls_session_t  and (as cookie verification etc. has taken place) this is a good time to set up a new session to hang off endpoint->sessions (and OpenSSL does the same thing after doing DTLSv1_listen() - happens to be the same point in the SSL handshake and then goes on to complete the handshakes with SSL_accept()).

As gnutls_handshake() and hence do_gnutls_handshake(), continues to respond with a negative error right up to the point the SSL handshake successfully completes, the code above is doing the right thing be sending back 1 if the client_hello has been seen, even though do_gnutls_handshake() returns -1 - as there is still the server hello, etc. to go, but that is safe to do as a separate session away from the endpoint->hello.

Reply:
Reworked the code so that `seen_client_hello` is always cleared when exiting from coap_dtls_hello() with a return code of 1 - which causes the proper coap_session to get set up.